data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java

```diff
@@ -141,8 +144,18 @@ private Map<String, List<Record<Event>>> groupRecordsBasedOnIdentificationKeys(
             final Event event = record.getData();
 
             final List<String> identificationKeyValues = new LinkedList<>();
+            int numMissingIdentificationKeys = 0;
             for (final String identificationKey : identificationKeys) {
-                identificationKeyValues.add(event.get(identificationKey, Object.class).toString());
+                final Object identificationKeyValue = event.get(identificationKey, Object.class);
+                if (identificationKeyValue == null) {
+                    identificationKeyValues.add(null);
+                    numMissingIdentificationKeys++;
+                } else {
+                    identificationKeyValues.add(identificationKeyValue.toString());
+                }
+            }
+            if (numMissingIdentificationKeys == identificationKeys.size()) {
+                recordsMissingIdentificationKeys.increment(1);
```

Comment:
Should we now clear the `identificationKeys`? As it is, when all the keys are missing we will create a hash from a non-empty list of all null values. These would then all be hashed to a single node in the cluster.

It seems the best behavior is to keep them on the local node. Clearing `identificationKeys` will keep this on the local endpoint.

https://github.com/opensearch-project/data-prepper/blob/87ba9cdd4198b41da79547d9231a6135fb0bfd1a/data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/HashRing.java#L53-L55